scripts/brp-python-bytecompile

```diff
@@ -1,34 +1,110 @@
-       xargs -0 -n 1 dirname | sed 's,[^/],,g' | sort -u | tail -n 1 | wc -c`
-if [ -z "$depth" ] || [ "$depth" -le "1" ]; then
-	exit 0
-fi
+# This function clamps the source mtime, see https://fedoraproject.org/wiki/Changes/ReproducibleBuildsClampMtimes
+function python_clamp_source_mtime()
+{
+    local _=$1
+    local python_binary=$2
+    local _=$3
+    local python_libdir="$4"
+    PYTHONPATH=/usr/lib/rpm/redhat/ $python_binary -B -m clamp_source_mtime -q "$python_libdir"
+}
 
+# This function now implements Python byte-compilation in three different ways:
+# Python >= 3.4 and < 3.9 uses a new module compileall2 - https://github.com/fedora-python/compileall2
+# In Python >= 3.9, compileall2 was merged back to standard library (compileall) so we can use it directly again.
+# Python < 3.4 (inc. Python 2) uses compileall module from stdlib with some hacks
 function python_bytecompile()
 {
     local options=$1
     local python_binary=$2
     local exclude=$3
-    local python_libdir=$4
-    local depth=$5
-    local real_libdir=$6
+    local python_libdir="$4"
+    local compileall_flags="$5"
+
+    python_version=$($python_binary -c "import sys; sys.stdout.write('{0.major}{0.minor}'.format(sys.version_info))")
+
+    #
+    # Python 3.4 and higher
+    #
+    if [ "$python_version" -ge 34 ]; then
+
+        # We compile all opt levels in one go: only when $options is empty.
+        if [ -n "$options" ]; then
+            return
+        fi
+
+        if [ "$python_version" -ge 39 ]; then
+            # For Pyhon 3.9+, use the standard library
+            compileall_module=compileall
+        else
+            # For older Pythons, use compileall2
+            compileall_module=compileall2
+        fi
+
+        if [ "$python_version" -ge 37 ]; then
+            # Force the TIMESTAMP invalidation mode
+            invalidation_option=--invalidation-mode=timestamp
+        else
+            # For older Pythons, the option does not exist
+            # as the invalidation is always based on size+mtime
+            invalidation_option=
+        fi
+
+        [ ! -z $exclude ] && exclude="-x '$exclude'"
+
+        # PYTHONPATH is needed for compileall2, but doesn't hurt for the stdlib
+        # -o 0 -o 1 are the optimization levels
+        # -q disables verbose output
+        # -f forces the process to overwrite existing compiled files
+        # -x excludes paths defined by regex
+        # -e excludes symbolic links pointing outside the build root
+        # -x and -e together implements the same functionality as the Filter class below
+        # -s strips $RPM_BUILD_ROOT from the path
+        # -p prepends the leading slash to the path to make it absolute
+        PYTHONPATH=/usr/lib/rpm/redhat/ $python_binary -B -m $compileall_module $compileall_flags -o 0 -o 1 -q -f $exclude -s "$RPM_BUILD_ROOT" -p / --hardlink-dupes $invalidation_option -e "$RPM_BUILD_ROOT" "$python_libdir"
```

Comment:
We can't assume `/usr/lib/rpm/redhat`

Reply:
How about referring to a vendor-neutral location (from platform.in this time)?

Reply:
hm... actually, I probably need to also pull the compileall2 module from python-rpm-macros and possibly add it to b3063f3f or just after that commit.